src/cpp/src/llm_pipeline.cpp

```diff
@@ -65,6 +68,22 @@ std::pair<EncodedResults, int32_t> beam_search(
 );
 
 class StatefulLLMPipeline final : public LLMPipelineImplBase {
+private:
+    void slice_matmul_statefull_model(std::shared_ptr<ov::Model> model) {
+        auto last_node = model->output(0).get_node()->input_value(0).get_node();
+        if (auto matmul = dynamic_cast<ov::op::v0::MatMul*>(last_node)) {
+            auto shape = matmul->input(0).get_partial_shape();
+            if (shape.rank().get_length() == 3 && shape[1] != 1) {
+                auto start = std::make_shared<ov::op::v0::Constant>(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{-1});
+                auto stop = std::make_shared<ov::op::v0::Constant>(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{-2});
+                auto step = std::make_shared<ov::op::v0::Constant>(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{-1});
+                auto axis = std::make_shared<ov::op::v0::Constant>(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{1});
+                auto slice = std::make_shared<ov::op::v8::Slice>(matmul->input_value(0), start, stop, step, axis);
+                matmul->input(0).replace_source_output(slice);
+            }
```

Comment:
I think else brunch must throw so we don't need to guess why no improvement is observed.

Reply:
or we can cast and add OPENVINO_ASSERT that casted value is not nullptr

Reply:
done